piptools/utils.py

```diff
@@ -115,7 +115,18 @@ def format_requirement(
     if ireq.editable:
         line = f"-e {ireq.link.url}"
     elif is_url_requirement(ireq):
-        line = ireq.link.url
+        # if the requirement has no name then it's just a URL
+        if not ireq.name:
+            line = ireq.link.url
+        # if it starts with "file:", PEP508 does not support direct reference
+        elif ireq.link.url.startswith("file:"):
```

Comment:
> PEP508 does not support direct reference

I couldn't find any information about that in the PEP. I've checked the grammar and couldn't found any restrictions on `name @ file:...` too.

I've tried locally absolute paths and it works:

```console
$ pip install 'pip-tools @ file:///Users/albert/Projects/pip-tools'
...
Installing collected packages: pip-tools
  Attempting uninstall: pip-tools
    Found existing installation: pip-tools 6.1.1.dev72+g39856be
    Uninstalling pip-tools-6.1.1.dev72+g39856be:
      Successfully uninstalled pip-tools-6.1.1.dev72+g39856be
Successfully installed pip-tools-6.1.1.dev72+g39856be
```

Local path doesn't work though:

```
$ pip install 'pip-tools @ file:./'
ERROR: Invalid requirement: 'pip-tools @ file:./'

$ pip install 'pip-tools @ file:.'
ERROR: Invalid requirement: 'pip-tools @ file:.'

$ pip install 'pip-tools @ file:pip-tools'
ERROR: Invalid requirement: 'pip-tools @ file:pip-tools'
```

Reply:
Sorry about that, you're definitely right. There's even an example in pip's documentation about it: https://pip.pypa.io/en/stable/cli/pip_install/#requirement-specifiers

I have updated the code again. It looks simpler IMO now. Let me know what you think.
Also I'd love to add a test in `test_cli_compile.py` about a direct reference with a relative path in `file:` but it seems tricky to me because as far as I know the unit tests are working in a temp dir.

Reply:
@atugushev I have updated the tests in `test_cli_compile.py`.